watchtower/wtclient/addr_iterator.go

```diff
@@ -0,0 +1,268 @@
+// removing the addresses in question from the iterator.
+type AddressIterator struct {
+	mu             sync.Mutex
+	addrList       *list.List
+	currentTopAddr *list.Element
+	candidates     map[string]*candidateAddr
+	lockedCount    int
+}
+
+type candidateAddr struct {
+	addr   net.Addr
+	locked bool
+}
+
+// newAddressIterator constructs a new AddressIterator.
+func newAddressIterator(addrs ...net.Addr) (*AddressIterator, error) {
+	if len(addrs) == 0 {
+		return nil, fmt.Errorf("must have at least one address")
+	}
+
+	iter := &AddressIterator{
+		addrList:   list.New(),
+		candidates: make(map[string]*candidateAddr),
+	}
+
+	for _, addr := range addrs {
+		addrID := addr.String()
+		iter.addrList.PushBack(addrID)
+		iter.candidates[addrID] = &candidateAddr{addr: addr}
+	}
+	iter.Reset()
+
+	return iter, nil
+}
+
+// Reset clears the iterators state, and makes the address at the front of the
+// list the next item to be returned.
+func (a *AddressIterator) Reset() {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.unsafeReset()
+}
+
+// unsafeReset clears the iterator state and makes the address at the front of
+// the list the next item to be returned.
+//
+// NOTE: this method is not thread safe and so should only be called if the
+// appropriate mutex is being held.
+func (a *AddressIterator) unsafeReset() {
+	// Reset the next candidate to the front of the linked-list.
+	a.currentTopAddr = a.addrList.Front()
+}
+
+// Next returns the next candidate address. This iterator will always return
+// candidates in the order given when the iterator was instantiated. If no more
+// candidates are available, ErrAddressesExhausted is returned.
+func (a *AddressIterator) Next() (net.Addr, error) {
+	return a.next(false)
+}
+
+// NextAndLock does the same as described for Next, and it also places a lock on
+// the returned address so that the address can not be removed until the lock on
+// it has been released via ReleaseLock.
+func (a *AddressIterator) NextAndLock() (net.Addr, error) {
+	return a.next(true)
+}
+
+// next returns the next candidate address. This iterator will always return
+// candidates in the order given when the iterator was instantiated. If no more
+// candidates are available, ErrAddressesExhausted is returned.
+func (a *AddressIterator) next(lock bool) (net.Addr, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// Set the next candidate to the subsequent element.
+	a.currentTopAddr = a.currentTopAddr.Next()
+
+	for a.currentTopAddr != nil {
+		// Propose the address at the front of the list.
+		addrID := a.currentTopAddr.Value.(string)
+
+		// Check whether this address is still considered a candidate.
+		// If it's not, we'll proceed to the next.
+		candidate, ok := a.candidates[addrID]
+		if !ok {
+			nextCandidate := a.currentTopAddr.Next()
+			a.addrList.Remove(a.currentTopAddr)
+			a.currentTopAddr = nextCandidate
+			continue
+		}
+
+		if lock {
+			candidate.locked = true
+			a.lockedCount++
```

Comment:
I think we also need to check that `candidate.locked` isn't already `true` before increasing `a.lockedCount`.

Reply:
ah, yes of course. Silly me. 
Actually - this made me realise that we do actually want to allow multiple callers to place individual locks on addrs. And we dont want one caller to completely release a lock when another caller has not yet called `ReleaseLock`. So to accommodate for this, I have changed it a bit so that each candidateAddr has a "numLocks" (like a reference counter")

Reply:
Ok, awesome, great to see that the changed design also allows this use case!

Reply:
the only thing i can think of that will cause an issue is if a caller calls `ReleaseLock()` more than once accidentally... 
Do you think this is something we need to take into account or can we just be explicit that it is the callers responsibility to use the API properly? 

Otherwise, perhaps what we should do is let the `*Lock` methods return some kind of use-once-only call back that they can use to release the lock that they have placed thoughts?

Reply:
I think the current solution of having lock counts for individual candidate addresses is good as addresses having multiple locks on one candidate perfectly.